nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/listen/dispatcher/DatagramChannelDispatcher.java

```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    @Override
+    public void open(final int port, int maxBufferSize) throws IOException {
+        datagramChannel = DatagramChannel.open();
+        datagramChannel.configureBlocking(false);
+        if (maxBufferSize > 0) {
+            datagramChannel.setOption(StandardSocketOptions.SO_RCVBUF, maxBufferSize);
+            final int actualReceiveBufSize = datagramChannel.getOption(StandardSocketOptions.SO_RCVBUF);
+            if (actualReceiveBufSize < maxBufferSize) {
+                logger.warn("Attempted to set Socket Buffer Size to " + maxBufferSize + " bytes but could only set to "
+                        + actualReceiveBufSize + "bytes. You may want to consider changing the Operating System's "
+                        + "maximum receive buffer");
+            }
+        }
+        datagramChannel.socket().bind(new InetSocketAddress(port));
+        selector = Selector.open();
+        datagramChannel.register(selector, SelectionKey.OP_READ);
+    }
+
+    @Override
+    public void run() {
+        final ByteBuffer buffer = bufferPool.poll();
+        while (!stopped) {
+            try {
+                int selected = selector.select();
+                if (selected > 0){
+                    Iterator<SelectionKey> selectorKeys = selector.selectedKeys().iterator();
+                    while (selectorKeys.hasNext()) {
+                        SelectionKey key = selectorKeys.next();
+                        selectorKeys.remove();
+                        if (!key.isValid()) {
+                            continue;
+                        }
+                        DatagramChannel channel = (DatagramChannel) key.channel();
+                        SocketAddress socketAddress;
+                        buffer.clear();
+                        while (!stopped && (socketAddress = channel.receive(buffer)) != null) {
+                            String sender = "";
+                            if (socketAddress instanceof InetSocketAddress) {
+                                sender = ((InetSocketAddress) socketAddress).getAddress().toString();
+                            }
+
+                            // create a byte array from the buffer
+                            buffer.flip();
+                            byte bytes[] = new byte[buffer.limit()];
+                            buffer.get(bytes, 0, buffer.limit());
+
+                            final Map<String,String> metadata = EventFactoryUtil.createMapWithSender(sender);
+                            final E event = eventFactory.create(bytes, metadata, null);
+
+                            // queue the raw message with the sender, block until space is available
+                            events.put(event);
+                            buffer.clear();
+                        }
+                    }
+                }
+            } catch (InterruptedException e) {
+                stopped = true;
```

Comment:
when reviewing one of @olegz's patches, I thought this was a place where adding a Thread.currentThread().interrupt() made sense